### DISCUSSION TITLE
Metric events - what is the algorithm of summary metrics?

### DISCUSSION THREAD

**Discussion by AlexSTAL1980** [vectordotdev/vector#21556] [Q&A]
Hi!

Quote from the page [Metric events](https://vector.dev/docs/about/under-the-hood/architecture/data-model/metric/#summary):
_Similar to a histogram, a summary samples observations (usually things like request durations and response sizes). While it also provides a total count of observations and a sum of all observed values, **it calculates configurable quantiles over a sliding time window**._

What is the algorithm for "sliding time window"?
By a certain number of metrics? Or for a certain time period? Or some other algorithm?

I tried to analyze the source code and found the "entry point" in file \vector-master\src\sinks\prometheus\collector.rs
```
                MetricValue::Distribution {
                    samples,
                    statistic: StatisticKind::Summary,
                } => {
                    if let Some(statistic) = DistributionStatistic::from_samples(samples, quantiles)
```
where quantiles are calculated. But I don't have enough experience to understand how and where the "samples" array is formed.

Please tell me the complete sampling algorithm.

**Comment by jszwedko** [comment]
Hey! When converting a Prometheus distribution into an internal Vector "Summary" metric, there are no samples. There are just the summary statistics: count, sum, min, max, and quantiles. See https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md#histograms-and-summaries for details on how Prometheus serializes histograms.

**Comment by AlexSTAL1980** [comment]
Hey! Did I understand correctly from the answer that the above-mentioned documentation Vector "Metric events" contains an error, and there is no "**sliding time window**", and the statistics are calculated for the entire available period?

If I understood correctly, then in my opinion this is incorrect, and quantiles for export to Prometheus must be calculated in the vector for a configurable time interval or for a specified number of values.

I will provide a translation of an excerpt from [the article](https://habr.com/ru/companies/tochka/articles/690814/):

Usually for metrics we sample data, that is, from the entire set we select only some elements in order to calculate percentiles and other properties from them. How to select a finite number of elements from an infinite series of data? You will have to sacrifice something, there are different algorithms for this. We store, **for example, up to 100 elements**, and when we need to write a new one, we decide which element to throw out:
* **Sliding Window**: just throw out the oldest element. Very easy to implement and understand
* Forward Decay: throw away a random element, but the older it is, the more likely it is to be thrown away. The probability can grow in any way, for example, exponentially. The idea is that old values ​​still slightly affect the properties of the array, so this sample describes our infinite series of data a little more accurately.

Otherwise, it turns out that Max always increases (and never decreases), Min always decreases (and never increases), and the quantiles themselves are always averaged over the entire available range.

And one of the purposes of this type of metrics is a visual representation of the query execution time, which should change quickly (for example, over the last 5 minutes, but not for the entire available history).

**Reply by jszwedko** [reply]
Summary statistics do have a timestamp. That's a good question about whether they are expected to represent the entire period or just a window. Metrics can be either "incremental" or "absolute" so perhaps an "incremental summary" would just represent the time window since the last summary metric.

For Prometheus, though, I believe the summaries are absolute (that is: they represent a summary of the entire range).

**Comment by AlexSTAL1980** [comment]
Please tell me, why do we need to know the "average temperature in the hospital"?
I found a good technical article: [How summary metrics work in Prometheus](https://grafana.com/blog/2022/03/01/how-summary-metrics-work-in-prometheus/).

Main thoughts:

* Prometheus has its own client, in the example in Java: 
Summary metrics are implemented in the Prometheus client libraries, like client_golang or client_java. You call the library in your REST application, and the library will calculate the percentiles for you.

* a sliding time window is set in the client:
**You don’t want to get latency metrics for the entire runtime of an application. You want to look at a reasonable time interval.** For example, client_java implements a configurable sliding window which **is 10 minutes by default**.

That is, when calculating the summary metrics, there must be one more **parameter - a sliding time window**, as it is implemented in the same Prmetheus client.

Otherwise it doesn't make any practical sense.

**Reply by jszwedko** [reply]
Gotcha, I see. https://prometheus.io/docs/concepts/metric_types/#summary does seem to indicate that the quantiles are meant to be over some time range. That time range isn't encoded in the metric itself, though, which just has a singular timestamp. It depends on the client configuration as mentioned on https://prometheus.io/docs/practices/histograms/. So you need to know to interpret the summary as a "summary of the last 10 minutes" when querying the data.

As an example, this is what a summary looks like in the Prometheus exposition format:

```
# Finally a summary, which has a complex representation, too:
            # HELP rpc_duration_seconds A summary of the RPC duration in seconds.
            # TYPE rpc_duration_seconds summary
            rpc_duration_seconds{quantile="0.01"} 3102
            rpc_duration_seconds{quantile="0.05"} 3272
            rpc_duration_seconds{quantile="0.5"} 4773
            rpc_duration_seconds{quantile="0.9"} 9001
            rpc_duration_seconds{quantile="0.99"} 76656
            rpc_duration_seconds_sum 1.7560473e+07
            rpc_duration_seconds_count 4.588206224e+09
```

**Reply by AlexSTAL1980** [reply]
Requesting data will not help. This calculation can only be implemented on the client, as it written everywhere [VictoriaMetrics Key concepts](https://docs.victoriametrics.com/keyconcepts/#summary):
_Summary metric type is quite similar to histogram and is used for quantiles calculations. The main difference is that **calculations are made on the client-side**, so metrics exposition format already contains pre-defined quantiles_

And again from there:
_It is impossible to calculate quantiles for measurements collected over an arbitrary time range. Usually, summary quantiles **are calculated over a fixed time range such as the last 5 minutes**._

The sliding time window must be added to the parameter and the calculation must be implemented on the client (vector.dev)

**Reply by jszwedko** [reply]
I feel like I've lost what you are trying to do here 😄 Can you describe your use-case?

**Reply by AlexSTAL1980** [reply]
Unfortunately, I am not familiar with the Rust programming language and cannot understand in the source code where and by what algorithm the summary metric is calculated. According to the links I provided above, there are at least 3 algorithms for calculation.

Most likely, for a full implementation of the summary metrics, a separate transformer will be required, which will have a "sliding_time_window" parameter, specified in seconds or minutes, where the calculation will take place (now there is a similar "Aggregate", which has a time window for calculating "interval_ms").

That is, it is necessary to implement a full calculation of the summary metrics in Vector

The current implementation is wrong

**Reply by AlexSTAL1980** [reply]
Addition
And the usage scenario itself is simple - monitoring the duration of requests, response sizes, etc.

**Reply by jszwedko** [reply]
I see, yes, I think what you might be asking for is the ability to calculate summary metrics.

**Reply by AlexSTAL1980** [reply]
Unfortunately, the language barrier prevents me from understanding your answer accurately.

**Comment by KimMJ** [comment]
hello. I too use log-to-metric and I believe the histogram needs a sliding time window. I wanted to get a latency histogram for a specific api from nginx logs, which can be implemented via bucket, but also via summary type. I want to use the summary type to reduce the load on the metrics system, and simply see the p95 stats for each pod. So I initially tried to use the summary type in vector, but after realizing that it gives stats for the entire time period, I ended up using the method via bucket, which is semantically more correct, but also has the problem of creating too many series in some cases. Perhaps people who use the summary type would like to get stats over a period of time, and a feature toggle would be a good way to give them a choice.
